codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -692,9 +665,185 @@ void decrementFlowControlWindow(int bytes) {
          * Discards this {@link FlowControlled}, writing an error. If this frame is in the pending queue,
          * the unwritten bytes are removed from this branch of the priority tree.
          */
-        void writeError(FlowControlled frame, Http2Exception cause) {
+        private void writeError(FlowControlled frame, Http2Exception cause) {
             decrementPendingBytes(frame.size());
             frame.error(cause);
         }
     }
+
+    /**
+     * The remote flow control state for a single stream that is not in a state where flow controlled frames cannot
+     * be exchanged.
+     */
+    private static final class ReducedFlowState extends RemoteFlowControlState {
```

Comment:
@buchgr - Agreed there is more code, but the code is not complex.  Lets continue the rest of the discussion https://github.com/netty/netty/pull/3616#discussion_r28702880.  Once the trigger points are refined such that we don't need duplicate allocations we will be in better shape.